src/oic/oic/provider.py

```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+            _info = self.do_back_channel_logout(
+                self.cdb[client_id], _subject_id, session_id
+            )
+            if _info:
+                logout_spec["back_channel"] = {client_id: _info}
+        elif "frontchannel_logout_uri" in self.cdb[client_id]:
+            # Construct an IFrame
+            _iframe = self.do_front_channel_logout_iframe(
+                self.cdb[client_id], self.name, session_id
+            )
+            if _iframe:
+                logout_spec["front_channel"] = {client_id: _iframe}
+
+        return logout_spec
+
+    def end_session_endpoint(
+        self,
+        request: str = "",
+        cookie: Optional[Union[str, SimpleCookie]] = None,
+        **kwargs
+    ) -> Response:
+        """
+        Handle a RP initiated Logout request.
+
+        :param request: The logout request
+        :param cookie:
+        :param kwargs:
+        :return: Returns a dictionary with one key 'sjwt' and the value
+            being a signed JWT token with session information.
+        """
+        _req = self.server.message_factory.get_request_type("endsession_endpoint")
+        esr = _req().from_urlencoded(request)
+
+        logger.debug("End session request: %s", sanitize(esr.to_dict()))
+
+        if self.events:
+            self.events.store("protocol request", esr)
+
+        # 2 ways of find out client ID and user. Either through a cookie
+        # or using the id_token_hint. If I get information from both make sure they match
+        _, client_id, uid = self._get_uid_from_cookie(cookie)
+
+        sid = ""
+
+        if "id_token_hint" in esr:
+            id_token_hint = IdToken().from_jwt(
+                esr["id_token_hint"], keyjar=self.keyjar, verify=True
+            )
+            far_away = 86400 * 30  # 30 days
+
+            if client_id:
+                args = {"client_id": client_id}
+            else:
+                args = {}
+
+            try:
+                id_token_hint.verify(
+                    iss=self.baseurl, skew=far_away, nonce_storage_time=far_away, **args
+                )
+            except (VerificationError, NotForMe) as err:
+                logger.warning("Verification error on id_token_hint: %s", err)
+                return error_response("invalid_request", "Bad Id Token hint")
+
+            sub = id_token_hint["sub"]
+
+            if uid is not None:
+                # verify that 'sub' are bound to 'uid'
+                if self.sdb.get_uid_by_sub(sub) != uid:
+                    return error_response("invalid_request", "Wrong user")
+            else:
+                uid = self.sdb.get_uid_by_sub(sub)
+
+            if client_id is None:
+                if len(id_token_hint["aud"]) == 1:
+                    client_id = id_token_hint["aud"][0]
+                else:
+                    client_id = id_token_hint["azp"]
+
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
```

Comment:
What happens if no match is found? Also that case is not covered in tests.

Reply:
OK